File: 백준/Gold/10026. 적록색약/적록색약.cc
#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

int N;
char grid[100][100];
bool visited[100][100];
int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};

void dfs(int x, int y, char color, bool colorBlind) {
    visited[x][y] = true;
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (nx >= 0 && ny >= 0 && nx < N && ny < N && !visited[nx][ny]) {
            if (colorBlind) {
                if ((color == 'R' || color == 'G') && (grid[nx][ny] == 'R' || grid[nx][ny] == 'G')) {
                    dfs(nx, ny, color, colorBlind);
                } else if (grid[nx][ny] == color) {
                    dfs(nx, ny, color, colorBlind);
                }
            } else {
                if (grid[nx][ny] == color) {
                    dfs(nx, ny, color, colorBlind);
                }
            }
        }
    }
}

int main() {
    cin >> N;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> grid[i][j];
        }
    }

    int normalCount = 0;
    memset(visited, false, sizeof(visited));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (!visited[i][j]) {
                dfs(i, j, grid[i][j], false);
                normalCount++;
            }
        }
    }

    int colorBlindCount = 0;
    memset(visited, false, sizeof(visited));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (!visited[i][j]) {
                dfs(i, j, grid[i][j], true);
                colorBlindCount++;
            }
        }
    }

    cout << normalCount << " " << colorBlindCount << endl;
    return 0;
}

File: 백준/Gold/10026. 적록색약/적록색약.py
import sys
sys.setrecursionlimit(10**6)

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

n = int(input())
grid = [list(input()) for _ in range(n)]
visited = []


def dfs(x, y, color):
  visited[x][y] = True
  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]
    if 0 <= nx < n and 0 <= ny < n and not visited[nx][ny] and grid[nx][
      ny] == color:
      dfs(nx, ny, color)


def cnt():
  global visited

  normal_count = 0
  blind_count = 0
  visited = [[False] * n for _ in range(n)]
  for i in range(n):
    for j in range(n):
      if not visited[i][j]:
        dfs(i, j, grid[i][j])
        normal_count += 1

  for i in range(n):
    for j in range(n):
      if grid[i][j] == 'R':
        grid[i][j] = 'G'

  visited = [[False] * n for _ in range(n)]
  for i in range(n):
    for j in range(n):
      if not visited[i][j]:
        dfs(i, j, grid[i][j])
        blind_count += 1

  return normal_count, blind_count


normal, blind = cnt()
print(normal, blind)
